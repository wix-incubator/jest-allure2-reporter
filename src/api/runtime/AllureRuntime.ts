import type { Metadata } from 'jest-metadata';
import type { LabelName, Status, StatusDetails } from 'jest-allure2-reporter';

import type { ParameterOrString } from './index';
import { isPromiseLike } from '../../utils/isPromiseLike';
import { wrapFunction } from '../../utils/wrapFunction';
import { formatString } from '../../utils/formatString';

import type {
  AllureRuntimeBindOptions,
  IAllureRuntime,
  ParameterOptions,
} from './IAllureRuntime';
import { AllureTestCaseMetadataProxy } from './proxies';

export type AllureRuntimeConfig = {
  metadataProvider: () => Metadata;
  nowProvider: () => number;
};

const constant =
  <T>(value: T) =>
  () =>
    value;
const noop = (..._arguments: unknown[]) => void 0;

export class AllureRuntime implements IAllureRuntime {
  readonly #metadataProvider: AllureRuntimeConfig['metadataProvider'];
  readonly #now: AllureRuntimeConfig['nowProvider'];
  #idle = Promise.resolve();

  get #metadata(): AllureTestCaseMetadataProxy {
    return new AllureTestCaseMetadataProxy(this.#metadataProvider());
  }

  constructor(config: AllureRuntimeConfig) {
    this.#metadataProvider = config.metadataProvider;
    this.#now = config.nowProvider;
  }

  $bind(options?: AllureRuntimeBindOptions): AllureRuntime {
    const { metadata = true, time = false } = options ?? {};

    return new AllureRuntime({
      metadataProvider: metadata
        ? constant(this.#metadataProvider())
        : this.#metadataProvider,
      nowProvider: time ? constant(this.#now()) : this.#now,
    });
  }

  $plug(_callback: any): any {
    // TODO: implement
  }

  async flush(): Promise<void> {
    await this.#idle;
  }

  description(value: string) {
    this.#metadata.push('description', [value]);
  }

  descriptionHtml(value: string) {
    this.#metadata.push('descriptionHtml', [value]);
  }

  label(name: LabelName | string, value: string) {
    this.#metadata.push('labels', [{ name, value }]);
  }

  link(url: string, name = url, type?: string) {
    this.#metadata.push('links', [{ name, url, type }]);
  }

  parameter(name: string, value: unknown, options?: ParameterOptions) {
    this.#metadata.push('parameters', [
      {
        name,
        value: `${value}`,
        ...options,
      },
    ]);
  }

  parameters(parameters: Record<string, unknown>) {
    for (const [name, value] of Object.entries(parameters)) {
      this.parameter(name, value);
    }
  }

  status(status: Status, statusDetails?: StatusDetails) {
    if (statusDetails) {
      this.#metadata.assign({ status, statusDetails });
    } else {
      this.#metadata.assign({ statusDetails });
    }
  }

  statusDetails(statusDetails: StatusDetails) {
    this.#metadata.set('statusDetails', statusDetails);
  }

  step<T = unknown>(name: string, function_: () => T): T {
    this.#startStep(name, function_);
    const end = this.#stopStep;

    let result: T;
    try {
      result = function_();

      if (isPromiseLike(result)) {
        this.#metadata.set('stage', 'running');

        result.then(
          () => end('passed'),
          (error) =>
            end('failed', { message: error.message, trace: error.stack }),
        );
      } else {
        end('passed');
      }

      return result;
    } catch (error: unknown) {
      end('failed', {
        message: (error as Error).message,
        trace: (error as Error).stack,
      });
      throw error;
    }
  }

  createStep<F extends Function>(
    nameFormat: string,
    maybeParameters: F | ParameterOrString[],
    maybeFunction?: F,
  ): F {
    const function_: any = maybeFunction ?? (maybeParameters as F);
    if (typeof function_ !== 'function') {
      throw new TypeError(
        `Expected a function, got ${typeof function_} instead`,
      );
    }

    const runtime = this;
    const userParameters = Array.isArray(maybeParameters)
      ? maybeParameters
      : null;

    const handleArguments = (arguments_: IArguments) => {
      const parameters = userParameters ?? Array.from(arguments_, noop);
      const allureParameters = parameters.map(resolveParameter, arguments_);
      for (const [name, value, options] of allureParameters) {
        runtime.parameter(name, value, options);
      }
    };

    return wrapFunction(function_, function (this: unknown) {
      const arguments_ = arguments;
      const name = formatString(nameFormat, ...arguments_);
      return runtime.step(
        name,
        wrapFunction(
          function_,
          function step(this: unknown) {
            handleArguments(arguments_);
            return Reflect.apply(function_, this, arguments_);
          }.bind(this),
        ),
      );
    });
  }

  #startStep = (name: string, function_: Function) => {
    this.#metadata.$startStep().assign({
      stage: 'scheduled',
      start: this.#now(),
      code: function_.toString(),
      description: [name],
    });
  };

  #stopStep = (status: Status, statusDetails?: StatusDetails) => {
    const existing = this.#metadata.get(undefined, {} as any);
    this.#metadata
      .assign({
        stage: 'finished',
        status: existing.status ?? status,
        statusDetails: existing.statusDetails ?? statusDetails,
        stop: this.#now(),
      })
      .$stopStep();
  };
}

function resolveParameter(
  this: unknown[],
  parameter: ParameterOrString | undefined,
  index: number,
) {
  const { name = `${index}`, ...options } =
    typeof parameter === 'string' ? { name: parameter } : parameter ?? {};

  return [name, this[index], options] as const;
}
