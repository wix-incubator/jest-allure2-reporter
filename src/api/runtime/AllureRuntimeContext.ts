import path from 'node:path';

import {
  StepsModule,
  ContentAttachmentsModule,
  CoreModule,
  FileAttachmentsModule,
} from './modules';
import * as attachmentHandlers from './attachment-handlers';
import type { MIMEInferer, SharedReporterConfig } from './types';
import type { AllureTestCaseMetadataProxy } from './proxies';

export type AllureRuntimeConfig = SharedReporterConfig & {
  inferMimeType: MIMEInferer;
  metadataProvider: () => AllureTestCaseMetadataProxy;
  nowProvider: () => number;
};

//
// /**
//  * @internal
//  */
// export interface AllureRuntimeContext {
//   readonly contentAttachmentHandlers: Record<string, ContentAttachmentHandler>;
//   readonly fileAttachmentHandlers: Record<string, FileAttachmentHandler>;
//   readonly metadata: AllureTestCaseMetadataProxy;
//   readonly now: Number;
//   idle: Promise<unknown>;
//   inferMimeType: MIMEInferer;
// }

export class AllureRuntimeContext {
  readonly modules: {
    basicSteps: StepsModule;
    core: CoreModule;
    contentAttachments: ContentAttachmentsModule;
    fileAttachments: FileAttachmentsModule;
  };

  readonly idle: Promise<unknown> = Promise.resolve();

  constructor(readonly config: AllureRuntimeConfig) {
    this.modules = {
      basicSteps: new StepsModule({
        get metadata() {
          return config.metadataProvider();
        },
        get now() {
          return config.nowProvider();
        },
      }),
      core: new CoreModule({
        get metadata() {
          return config.metadataProvider();
        },
      }),
      contentAttachments: new ContentAttachmentsModule({
        handlers: {
          write: attachmentHandlers.writeHandler,
        },
        inferMimeType: config.inferMimeType,
        get metadata() {
          return config.metadataProvider();
        },
        outDir: path.join(config.resultsDir, config.attachments.subDir),
        waitFor: (promise) => {
          this.idle.then(() => promise);
        },
      }),
      fileAttachments: new FileAttachmentsModule({
        handlers: {
          copy: attachmentHandlers.copyHandler,
          move: attachmentHandlers.moveHandler,
          ref: attachmentHandlers.referenceHandler,
        },
        inferMimeType: config.inferMimeType,
        get metadata() {
          return config.metadataProvider();
        },
        outDir: path.join(config.resultsDir, config.attachments.subDir),
        waitFor: (promise) => {
          this.idle.then(() => promise);
        },
      }),
    };
  }
}
