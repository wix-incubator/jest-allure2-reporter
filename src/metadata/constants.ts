export const PREFIX = 'allure2' as const;

export const SHARED_CONFIG = [PREFIX, 'config'] as const;

export const CODE = [PREFIX, 'code'] as const;
export const WORKER_ID = [PREFIX, 'workerId'] as const;
export const HIDDEN = [PREFIX, 'hidden'] as const;

export const START = [PREFIX, 'start'] as const;

export const STOP = [PREFIX, 'stop'] as const;

export const CURRENT_STEP = [PREFIX, 'currentStep'] as const;
export const DESCRIPTION = [PREFIX, 'description'] as const;
export const DESCRIPTION_HTML = [PREFIX, 'descriptionHtml'] as const;
export const LABELS = [PREFIX, 'labels'] as const;
export const LINKS = [PREFIX, 'links'] as const;
