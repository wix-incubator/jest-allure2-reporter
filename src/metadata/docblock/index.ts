// import {splitDocblock} from "../../utils";
// import type {
//   AllureTestCaseMetadata,
//   AllureTestFileMetadata,
//   AllureTestItemMetadata, AllureTestStepMetadata,
//   DocblockContext, Label, LabelName, Link, LinkType
// } from "jest-allure2-reporter";
//
// async function initParser(): Promise<DocblockParser> {
//   try {
//     const jestDocblock = await import('jest-docblock');
//     return (snippet) => {
//       const [jsdoc] = splitDocblock(snippet);
//       const result = jestDocblock.parseWithComments(jsdoc);
//       return {
//         raw: jsdoc,
//         comments: result.comments,
//         pragmas: normalize(result.pragmas),
//       };
//     };
//   } catch (error: any) {
//     // TODO: log warning
//     if (error?.code === 'MODULE_NOT_FOUND') {
//       return () => void 0;
//     }
//
//     throw error;
//   }
// }
//
// const SPLITTERS: Record<string, (string_: string) => string[]> = {
//   tag: (string_) => string_.split(/\s*,\s*/),
// };
//
// function normalize(
//   pragmas: Record<string, string | string[]>,
// ): Record<string, string[]> {
//   const result: Record<string, string[]> = {};
//
//   for (const [key, value] of Object.entries(pragmas)) {
//     result[key] = Array.isArray(value) ? value : [value];
//     const splitter = SPLITTERS[key];
//     if (splitter) {
//       result[key] = result[key].flatMap(splitter);
//     }
//   }
//
//   return result;
// }
//
// function hasDocblockAtStart(string_: string) {
//   return /^\s*\/\*\*/.test(string_);
// }
//
// function mergeIntoTestItem(
//   metadata: AllureTestItemMetadata,
//   comments: string,
//   pragmas: Record<string, string[]>,
//   rawDocblock: string,
//   shouldLeaveComments: boolean,
// ) {
//   if (comments) {
//     metadata.description ??= [];
//     metadata.description.push(comments);
//   }
//
//   if (pragmas.description) {
//     metadata.description ??= [];
//     metadata.description.push(...pragmas.description);
//   }
//
//   if (metadata.sourceCode && rawDocblock) {
//     const [left, right, ...rest] = metadata.sourceCode.split(rawDocblock);
//     const leftTrimmed = left.trimEnd();
//     const replacement = shouldLeaveComments
//       ? `/** ${comments.trimStart()} */\n`
//       : '\n';
//     const joined = right ? [leftTrimmed, right].join(replacement) : leftTrimmed;
//     metadata.sourceCode = [joined, ...rest].join('\n');
//   }
// }
//
// function mergeIntoTestFile(
//   metadata: AllureTestFileMetadata,
//   docblock: DocblockContext | undefined,
// ) {
//   return mergeIntoTestCase(metadata, docblock);
// }
//
// function mergeIntoTestCase(
//   metadata: AllureTestCaseMetadata,
//   docblock: DocblockContext | undefined,
// ) {
//   const { raw = '', comments = '', pragmas = {} } = docblock ?? {};
//   mergeIntoTestItem(metadata, comments, pragmas, raw, false);
//
//   const epic = pragmas.epic?.map(createLabelMapper('epic')) ?? [];
//   const feature = pragmas.feature?.map(createLabelMapper('feature')) ?? [];
//   const owner = pragmas.owner?.map(createLabelMapper('owner')) ?? [];
//   const severity = pragmas.severity?.map(createLabelMapper('severity')) ?? [];
//   const story = pragmas.story?.map(createLabelMapper('story')) ?? [];
//   const tag = pragmas.tag?.map(createLabelMapper('tag')) ?? [];
//   const labels = [...epic, ...feature, ...owner, ...severity, ...story, ...tag];
//   if (labels.length > 0) {
//     metadata.labels ??= [];
//     metadata.labels.push(...labels);
//   }
//
//   const issue = pragmas.issue?.map(createLinkMapper('issue')) ?? [];
//   const tms = pragmas.tms?.map(createLinkMapper('tms')) ?? [];
//   const links = [...issue, ...tms];
//   if (links.length > 0) {
//     metadata.links ??= [];
//     metadata.links.push(...links);
//   }
//
//   if (pragmas.descriptionHtml) {
//     metadata.descriptionHtml ??= [];
//     metadata.descriptionHtml.push(...pragmas.descriptionHtml);
//   }
// }
//
// function createLabelMapper(name: LabelName) {
//   return (value: string): Label => ({ name, value });
// }
//
// function createLinkMapper(type?: LinkType) {
//   return (url: string): Link => ({ type, url, name: url });
// }
//
// function mergeIntoTestStep(
//   metadata: AllureTestStepMetadata,
//   docblock: DocblockContext | undefined,
// ) {
//   const { raw = '', comments = '', pragmas = {} } = docblock ?? {};
//   mergeIntoTestItem(metadata, comments, pragmas, raw, true);
// }
//
// type DocblockParser = (raw: string) => DocblockContext | undefined;
//
