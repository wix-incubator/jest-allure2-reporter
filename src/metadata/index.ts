export * from './proxies';
export * from './MetadataSquasher';
