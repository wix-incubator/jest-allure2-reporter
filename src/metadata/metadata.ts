import type {
  Attachment,
  Label,
  Link,
  Parameter,
  Stage,
  Status,
  StatusDetails,
} from '@noomorph/allure-js-commons';

export interface AllureTestStepMetadata {
  steps?: AllureTestStepMetadata[];

  name?: string;
  status?: Status;
  statusDetails?: StatusDetails;
  stage?: Stage;
  attachments?: Attachment[];
  parameters?: Parameter[];
  start?: number;
  stop?: number;
}

export interface AllureTestCaseMetadata extends AllureTestStepMetadata {
  /**
   * Pointer to the child step that is currently being added or executed.
   * @example ['steps', '0', 'steps', '0']
   * @internal
   */
  currentStep?: string[];
  /**
   * Jest worker ID.
   * @internal Used to generate unique thread names.
   * @see {import('@noomorph/allure-js-commons').LabelName.THREAD}
   */
  workerId?: string;
  /**
   * Source code of the test case, glued from all hooks and test function itself.
   */
  code?: string[];
  /**
   * Only steps can have names.
   */
  name?: never;
  description?: string[];
  descriptionHtml?: string[];
  labels?: Label[];
  links?: Link[];
}
