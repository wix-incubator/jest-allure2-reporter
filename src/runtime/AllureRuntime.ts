import path from 'node:path';

import type {
  LabelName,
  ParameterOptions,
  StatusDetails,
} from '@noomorph/allure-js-commons';
import { Stage, Status } from '@noomorph/allure-js-commons';
import type { Metadata } from 'jest-metadata';
import type {
  AllureTestStepMetadata,
  AttachmentContent,
  AttachmentOptions,
  IAllureRuntime,
  ParameterOrString,
} from 'jest-allure2-reporter';

import {
  CURRENT_STEP,
  DESCRIPTION,
  DESCRIPTION_HTML,
  LABELS,
  LINKS,
  PREFIX,
} from '../constants';
import { isPromiseLike } from '../utils/isPromiseLike';
import { inferMimeType } from '../utils/inferMimeType';
import { hijackFunction } from '../utils/hijackFunction';
import type { Function_, MaybePromise } from '../utils/types';
import { processMaybePromise } from '../utils/processMaybePromise';
import { wrapFunction } from '../utils/wrapFunction';
import { formatString } from '../utils/formatString';

import type { IAttachmentsHandler } from './AttachmentsHandler';

export type AllureRuntimeConfig = {
  attachmentsHandler: IAttachmentsHandler;
  metadataProvider: () => Metadata;
  nowProvider: () => number;
};

const noop = (..._arguments: unknown[]) => void 0;

export class AllureRuntime implements IAllureRuntime {
  readonly #attachmentsHandler: AllureRuntimeConfig['attachmentsHandler'];
  readonly #metadataProvider: AllureRuntimeConfig['metadataProvider'];
  readonly #now: AllureRuntimeConfig['nowProvider'];
  #idle = Promise.resolve();

  get #metadata(): Metadata {
    return this.#metadataProvider();
  }

  constructor(config: AllureRuntimeConfig) {
    this.#attachmentsHandler = config.attachmentsHandler;
    this.#metadataProvider = config.metadataProvider;
    this.#now = config.nowProvider;
  }

  async flush(): Promise<void> {
    await this.#idle;
  }

  description(value: string) {
    this.#metadata.push(DESCRIPTION, [value]);
  }

  descriptionHtml(value: string) {
    this.#metadata.push(DESCRIPTION_HTML, [value]);
  }

  label(name: LabelName | string, value: string) {
    this.#metadata.push(LABELS, [{ name, value }]);
  }

  link(name: string, url: string, type?: string) {
    this.#metadata.push(LINKS, [{ name, url, type }]);
  }

  parameter(name: string, value: unknown, options?: ParameterOptions) {
    this.#metadata.push(this.#localPath('parameters'), [
      {
        name,
        value: `${value}`,
        ...options,
      },
    ]);
  }

  parameters(parameters: Record<string, unknown>) {
    for (const [name, value] of Object.entries(parameters)) {
      this.parameter(name, value);
    }
  }

  status(status?: Status | StatusDetails, statusDetails?: StatusDetails) {
    this.#metadata.assign(this.#localPath(), { status, statusDetails });
  }

  statusDetails(statusDetails: StatusDetails) {
    this.#metadata.assign(this.#localPath(), { statusDetails });
  }

  attachment<T extends AttachmentContent>(
    name: string,
    content: MaybePromise<T>,
    mimeType?: string,
  ): typeof content {
    return processMaybePromise(
      content,
      this.#handleDynamicAttachment({ name, mimeType }),
    );
  }

  createAttachment<T extends AttachmentContent>(
    function_: Function_<MaybePromise<T>>,
    rawOptions: string | AttachmentOptions,
  ): typeof function_ {
    const options = this.#resolveAttachmentOptions(rawOptions);
    return hijackFunction(function_, this.#handleDynamicAttachment(options));
  }

  fileAttachment(
    filePathOrPromise: MaybePromise<string>,
    rawOptions?: string | AttachmentOptions,
  ): any {
    const options = this.#resolveAttachmentOptions(rawOptions);
    return processMaybePromise(
      filePathOrPromise,
      this.#handleStaticAttachment(options),
    );
  }

  createFileAttachment(
    function_: Function_<MaybePromise<string>>,
    rawOptions?: string | AttachmentOptions,
  ): typeof function_ {
    const options = this.#resolveAttachmentOptions(rawOptions);
    return hijackFunction<string>(
      function_,
      this.#handleStaticAttachment(options),
    );
  }

  step<T = unknown>(name: string, function_: () => T): T {
    this.#startStep(name, function_);
    const end = this.#stopStep;

    let result: T;
    try {
      result = function_();

      if (isPromiseLike(result)) {
        this.#updateStep(Stage.RUNNING);

        result.then(
          () => end(Status.PASSED),
          (error) =>
            end(Status.FAILED, { message: error.message, trace: error.stack }),
        );
      } else {
        end(Status.PASSED);
      }

      return result;
    } catch (error: unknown) {
      end(Status.FAILED, {
        message: (error as Error).message,
        trace: (error as Error).stack,
      });
      throw error;
    }
  }

  createStep<F extends Function>(
    nameFormat: string,
    maybeParameters: F | ParameterOrString[],
    maybeFunction?: F,
  ): F {
    const function_: any = maybeFunction ?? (maybeParameters as F);
    if (typeof function_ !== 'function') {
      throw new TypeError(
        `Expected a function, got ${typeof function_} instead`,
      );
    }

    const runtime = this;
    const userParameters = Array.isArray(maybeParameters)
      ? maybeParameters
      : null;

    const handleArguments = (arguments_: IArguments) => {
      const parameters = userParameters ?? Array.from(arguments_, noop);
      const allureParameters = parameters.map(resolveParameter, arguments_);
      for (const [name, value, options] of allureParameters) {
        runtime.parameter(name, value, options);
      }
    };

    return wrapFunction(function_, function (this: unknown) {
      const arguments_ = arguments;
      const name = formatString(nameFormat, ...arguments_);
      return runtime.step(
        name,
        wrapFunction(
          function_,
          function step(this: unknown) {
            handleArguments(arguments_);
            return Reflect.apply(function_, this, arguments_);
          }.bind(this),
        ),
      );
    });
  }

  #startStep = (name: string, function_: Function) => {
    const count = this.#metadata.get(this.#localPath('steps', 'length'), 0);
    this.#metadata.push(this.#localPath('steps'), [
      {
        name,
        stage: Stage.SCHEDULED,
        start: this.#now(),
        code: function_.toString(),
      },
    ]);
    // eslint-disable-next-line unicorn/no-array-push-push
    this.#metadata.push(CURRENT_STEP, ['steps', `${count}`]);
  };

  #stopStep = (status: Status, statusDetails?: StatusDetails) => {
    const existing = this.#metadata.get(this.#localPath(), {} as any);

    this.#metadata.assign(this.#localPath(), {
      stage: Stage.FINISHED,
      status: existing.status ?? status,
      statusDetails: existing.statusDetails ?? statusDetails,
      stop: this.#now(),
    });

    const currentStep = this.#metadata.get(CURRENT_STEP, []) as string[];
    this.#metadata.set(CURRENT_STEP, currentStep.slice(0, -2));
  };

  #updateStep = (stage: Stage) => {
    this.#metadata.set(this.#localPath('stage'), stage);
  };

  #localPath(key?: keyof AllureTestStepMetadata, ...innerKeys: string[]) {
    const stepPath = this.#metadata.get(CURRENT_STEP, []) as string[];
    const allKeys = key ? [key, ...innerKeys] : innerKeys;
    return [PREFIX, ...stepPath, ...allKeys];
  }

  #resolveAttachmentOptions(rawOptions?: string | AttachmentOptions) {
    return !rawOptions || typeof rawOptions === 'string'
      ? { name: rawOptions, mimeType: undefined }
      : rawOptions;
  }

  #handleDynamicAttachment =
    ({ name: nameFormat, mimeType }: AttachmentOptions) =>
    (content: AttachmentContent, arguments_?: unknown[]) => {
      const name = this.#formatName(nameFormat, arguments_);
      const source = this.#attachmentsHandler.placeAttachment(name, content);
      this.#metadata.push(this.#localPath('attachments'), [
        {
          name,
          source: path.resolve(source),
          type: mimeType ?? inferMimeType(name),
        },
      ]);
      const promise = this.#attachmentsHandler.writeAttachment(source, content);
      this.#idle = this.#idle.then(() => promise);
    };

  #handleStaticAttachment =
    ({ name: rawName, mimeType }: AttachmentOptions) =>
    (filePath: string, arguments_?: unknown[]) => {
      const name = this.#formatName(
        rawName ?? path.basename(filePath),
        arguments_,
      );
      const securedPath = this.#attachmentsHandler.placeAttachment(name);
      const result = this.#attachmentsHandler.secureAttachment(
        filePath,
        securedPath,
      );
      this.#metadata.push(this.#localPath('attachments'), [
        {
          name,
          source: path.resolve(result.destinationPath),
          type: mimeType ?? inferMimeType(name, filePath),
        },
      ]);
      if (result.promise) {
        this.#idle = this.#idle.then(() => result.promise);
      }
    };

  #formatName(nameFormat?: string, arguments_?: unknown[]) {
    return arguments_
      ? formatString(nameFormat ?? 'untitled', ...arguments_)
      : nameFormat ?? 'untitled';
  }
}

function resolveParameter(
  this: unknown[],
  parameter: ParameterOrString | undefined,
  index: number,
) {
  const { name = `${index}`, ...options } =
    typeof parameter === 'string' ? { name: parameter } : parameter ?? {};

  return [name, this[index], options] as const;
}
