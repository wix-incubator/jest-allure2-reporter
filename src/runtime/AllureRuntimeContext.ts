import type {
  AllureGlobalMetadata,
  AllureTestFileMetadata,
} from 'jest-allure2-reporter';

import { once } from '../utils';
import { AllureMetadataProxy, AllureTestItemMetadataProxy } from '../metadata';

import type { AllureRuntimeConfig } from './AllureRuntimeConfig';
import * as attachmentHandlers from './attachment-handlers';
import type {
  ContentAttachmentHandler,
  FileAttachmentHandler,
  MIMEInferer,
  SharedReporterConfig,
} from './types';

export class AllureRuntimeContext {
  readonly contentAttachmentHandlers: Record<string, ContentAttachmentHandler> =
    {
      write: attachmentHandlers.writeHandler,
    };
  readonly fileAttachmentHandlers: Record<string, FileAttachmentHandler> = {
    copy: attachmentHandlers.copyHandler,
    move: attachmentHandlers.moveHandler,
    ref: attachmentHandlers.referenceHandler,
  };
  inferMimeType: MIMEInferer = attachmentHandlers.inferMimeType;
  readonly getReporterConfig: () => SharedReporterConfig;
  readonly getFileMetadata: () => AllureMetadataProxy<AllureTestFileMetadata>;
  readonly getGlobalMetadata: () => AllureMetadataProxy<AllureGlobalMetadata>;
  readonly getCurrentMetadata: () => AllureTestItemMetadataProxy;
  readonly getNow: () => number;
  idle: Promise<unknown> = Promise.resolve();

  constructor(config: AllureRuntimeConfig) {
    this.getNow = config.getNow;
    this.getReporterConfig = once(config.getReporterConfig);
    this.getCurrentMetadata = () =>
      new AllureTestItemMetadataProxy(config.getCurrentMetadata());
    this.getFileMetadata = () =>
      new AllureMetadataProxy(config.getFileMetadata());
    this.getGlobalMetadata = () =>
      new AllureMetadataProxy(config.getGlobalMetadata());

    Object.defineProperty(this.contentAttachmentHandlers, 'default', {
      get: () => {
        const defaultName = this.getReporterConfig().attachments.contentHandler;
        return this.contentAttachmentHandlers[defaultName];
      },
    });

    Object.defineProperty(this.fileAttachmentHandlers, 'default', {
      get: () => {
        const defaultName = this.getReporterConfig().attachments.fileHandler;
        return this.fileAttachmentHandlers[defaultName];
      },
    });
  }
}
