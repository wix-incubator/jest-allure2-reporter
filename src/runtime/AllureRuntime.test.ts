import path from 'node:path';

import { state } from 'jest-metadata';

import { AllureMetadataProxy } from '../metadata';

import { AllureRuntime } from './AllureRuntime';
import type { SharedReporterConfig } from './types';
import { AllureRuntimeContext } from './AllureRuntimeContext';

describe('AllureRuntime', () => {
  it('should add attachments within the steps', async () => {
    let now = 0;

    const context = new AllureRuntimeContext({
      contentAttachmentHandlers: {
        write: (context) =>
          path.join(context.outDir, context.content.toString()),
      },
      getCurrentMetadata: () => state.currentMetadata,
      getFileMetadata: () => state.lastTestFile!,
      getGlobalMetadata: () => state,
      getNow: () => now++,
      getReporterConfig(): SharedReporterConfig {
        return {
          overwrite: true,
          resultsDir: '/tmp',
          injectGlobals: false,
          attachments: {
            subDir: '../attachments',
            contentHandler: 'write',
            fileHandler: 'ref',
          },
        };
      },
    });

    const runtime = new AllureRuntime(context);
    runtime.attachment('attachment1', Buffer.from('first'), 'text/plain');

    const innerStep3 = runtime.createStep(
      'inner step 3',
      async (message: string) => {
        runtime.attachment('attachment4', message, 'text/plain');

        const error = new Error('Async error');
        error.stack = 'Test stack';
        throw error;
      },
    );

    await runtime.step('outer step', async () => {
      try {
        runtime.step('inner step 1', () => {
          runtime.attachment('attachment2', 'second', 'text/plain');
          const error = new Error('Sync error');
          error.stack = 'Test stack';
          // Simulating Jest assertion error
          Object.assign(error, { matcherResult: undefined });
          throw error;
        });
      } catch {
        /* empty */
      }
      runtime.step('inner step 2', () => {
        /* empty */
      });
      runtime.attachment('attachment3', 'third', 'text/plain');
      await innerStep3('fourth').catch(() => {
        /* empty */
      });
    });
    runtime.attachment('attachment5', Buffer.from('fifth'), 'text/plain');
    await runtime.flush();
    expect(new AllureMetadataProxy(state).get()).toMatchSnapshot();
  });
});
