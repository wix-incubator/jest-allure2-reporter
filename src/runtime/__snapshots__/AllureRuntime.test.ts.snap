// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AllureRuntime should add attachments within the steps 1`] = `
{
  "attachments": [
    {
      "name": "attachment1",
      "source": "/attachments/first",
      "type": "text/plain",
    },
    {
      "name": "attachment5",
      "source": "/attachments/fifth",
      "type": "text/plain",
    },
  ],
  "currentStep": [],
  "steps": [
    {
      "attachments": [
        {
          "name": "attachment3",
          "source": "/attachments/third",
          "type": "text/plain",
        },
      ],
      "code": "async () => {
            try {
                runtime.step('inner step 1', () => {
                    runtime.createAttachment('attachment2', 'second', 'text/plain');
                    const error = new Error('Sync error');
                    error.stack = 'Test stack';
                    throw error;
                });
            }
            catch {
                /* empty */
            }
            runtime.step('inner step 2', () => {
                /* empty */
            });
            await runtime.createAttachment('attachment3', 'third', 'text/plain');
            await runtime
                .step('inner step 3', async () => {
                await runtime.createAttachment('attachment4', 'fourth', 'text/plain');
                const error = new Error('Async error');
                error.stack = 'Test stack';
                throw error;
            })
                .catch(() => {
                /* empty */
            });
        }",
      "name": "outer step",
      "stage": "finished",
      "start": 0,
      "status": "passed",
      "statusDetails": undefined,
      "steps": [
        {
          "attachments": [
            {
              "name": "attachment2",
              "source": "/attachments/second",
              "type": "text/plain",
            },
          ],
          "code": "() => {
                    runtime.createAttachment('attachment2', 'second', 'text/plain');
                    const error = new Error('Sync error');
                    error.stack = 'Test stack';
                    throw error;
                }",
          "name": "inner step 1",
          "stage": "finished",
          "start": 1,
          "status": "failed",
          "statusDetails": {
            "message": "Sync error",
            "trace": "Test stack",
          },
          "stop": 2,
        },
        {
          "code": "() => {
                /* empty */
            }",
          "name": "inner step 2",
          "stage": "finished",
          "start": 3,
          "status": "passed",
          "statusDetails": undefined,
          "stop": 4,
        },
        {
          "attachments": [
            {
              "name": "attachment4",
              "source": "/attachments/fourth",
              "type": "text/plain",
            },
          ],
          "code": "async () => {
                await runtime.createAttachment('attachment4', 'fourth', 'text/plain');
                const error = new Error('Async error');
                error.stack = 'Test stack';
                throw error;
            }",
          "name": "inner step 3",
          "stage": "finished",
          "start": 5,
          "status": "failed",
          "statusDetails": {
            "message": "Async error",
            "trace": "Test stack",
          },
          "stop": 6,
        },
      ],
      "stop": 7,
    },
  ],
}
`;
